File: src/libs/dijkstra.py
from libs.heap import Heap


class Dijkstra:
    def __init__(self, g):
        self.g = g
        self.heap = Heap()
        self.heap.push(self.g.nodes[self.g.index_initial], self.g.index_initial)

    def execute(self):
        while len(self.heap) > 0:
            u_index = self.heap.pop()

            u = self.g.nodes[u_index]

            # If the search reaches the final node,
            # stop the search
            if u_index == self.g.index_final:
                break

            neighbors = u.generate_next_moves(
                self.g.nodes, self.g.original_weights, self.g.lookup
            )
            for v_index, uv_cost in neighbors:
                v = self.g.nodes[v_index]
                if v.distance > u.distance + uv_cost:
                    v.distance = u.distance + uv_cost
                    v.pi = u # Don't use pi
                    
                    # Check if node[v_index] was already visited in O(1)
                    # by seeing if the config is already inserted in the lookup
                    try:
                        self.g.lookup[self.g.nodes[v_index]]
                    except KeyError:
                        self.heap.push(v, v_index)

File: src/libs/dijkstra.cpp
#include "dijkstra.hpp"

Dijkstra::Dijkstra(Graph *_graph)
{
    graph = _graph;
    heap.Push(graph->nodes[graph->index_initial].distance, graph->index_initial);
}

void Dijkstra::Execute()
{
    std::vector<int> visited_nodes;
    while (heap.Size() > 0)
    {
        int u_index = heap.Pop();

        Node u = graph->nodes[u_index];

        // If the search reaches the final node, stop the search
        if (u_index == graph->index_final)
        {
            break;
        }

        visited_nodes.push_back(u_index);

        auto neighbors = graph->GenerateNextMoves(u);

        for (auto neighbor : neighbors)
        {
            int v_index = neighbor.first;
            int uv_cost = neighbor.second;

            Node &v = graph->nodes[v_index];

            if (v.distance > u.distance + uv_cost)
            {
                v.distance = u.distance + uv_cost;

                // Check if node[v_index] was already visited by seeing if the config is already inserted in the vector
                auto it = std::find(visited_nodes.begin(), visited_nodes.end(), v_index);
                if (it == visited_nodes.end())
                {
                    heap.Push(v.distance, v_index);
                }
            }
        }
    }
}
